Add stable_id support for match restoration

hoodie opened the pull request:
- Implement match_stable_id and find_match_by_stable_id for Outline
- Generate stable_id as "text:depth" for outline items
- Implement stable_id lookups for Project Symbols
- Implement stable_id support in Icon Theme Selector and Theme Selector
- Use stable_id to preserve and restore manual selections in Picker

Closes #35440

Release Notes:

- Added stickiness to manually selected items in project symbols and outline

SomeoneToIgnore commented on crates/picker/src/picker.rs:
```diff
@@ -704,7 +758,32 @@ impl<D: PickerDelegate> Picker<D> {
+            } else {
+                // Item no longer in results, clear manual selection and reset to first item
+                self.manually_selected_stable_id = None;
+                let ix = 0.min(match_count.saturating_sub(1));
```
`0.min(any_other_usize)` is `0` or am I missing something?

Also, should we try to fall back to closest selected index instead, to avoid large list jumps?

hoodie replied:
about those large jumps. wasn't there also a ticket about the heuristic that selects the very first item when opening the picker? I can't find it right now. anyway, right now the picker already picks a "relatively random" item instead of "the closest to the cursor". maybe I find that ticket again some day

SomeoneToIgnore reviewed (commented):
Sorry for the long wait time: had had quite a rough time and then a company-wide holiday.

Looks really great when I've tried it, thank you so much for finding a way to improve things generically.

One concern besides the listed ones: we need to test it.
At least, a test to show that the stable sorting works, and, ideally, another one to cover

> Item no longer in results, clear manual selection and reset to first item

branch.

hoodie commented on crates/picker/src/picker.rs:
```diff
@@ -398,11 +415,43 @@ impl<D: PickerDelegate> Picker<D> {
+    ///
+    /// Use this for user-driven selections (keyboard navigation, mouse clicks) where you want
+    /// the user's choice to be maintained as they continue typing. For programmatic selections
+    /// that should not persist, use `set_selected_index` instead.
```
intuitively that should be fine, I was just a bit careful to not change the default in too many places at once, because I had no good overview of what needs to be tested

hoodie commented on crates/picker/src/picker.rs:
```diff
@@ -444,6 +493,11 @@ impl<D: PickerDelegate> Picker<D> {
         self.delegate.set_selected_index(ix, window, cx);
         let current_index = self.delegate.selected_index();
 
+        // Track manually selected item to preserve across match updates
```
actually not, it subtly breaks the behavior, but I had to test it to find that out. Seems like I should add a few tests though to make sure of it.

hoodie commented on crates/outline/src/outline.rs:
```diff
@@ -290,6 +290,22 @@ impl PickerDelegate for OutlineViewDelegate {
+    fn find_match_by_stable_id(&self, stable_id: &str) -> Option<usize> {
+        self.matches.iter().position(|mat| {
+            if let Some(outline_item) = self.outline.items.get(mat.candidate_id) {
+                format!("{}:{}", outline_item.text, outline_item.depth) == stable_id
```
Time flies. I looked into this, this would be a bit of a larger change, because you can't really have one struct for all places. I could conceive of a trait here and implement it for specific pickers as needed. What do you think?
@SomeoneToIgnore

hoodie committed:
Use StableId trait for picker item identity

PickerDelegate now uses a StableId associated type instead of String
for stable item identification. This enables structured, efficient
comparison and avoids string formatting for item identity. All picker
implementations updated to use StableId, with () or SharedString as
appropriate.

hoodie committed:
Add stable id support for picker items

Preserves manual selections across match updates by introducing
BranchStableId and implementing StableId for branch picker items.

hoodie committed:
Add tests

hoodie committed:
Add stable_id support for match restoration

- Implement match_stable_id and find_match_by_stable_id for Outline
- Generate stable_id as "text:depth" for outline items
- Implement stable_id lookups for Project Symbols
- Implement stable_id support in Icon Theme Selector and Theme Selector
- Use stable_id to preserve and restore manual selections in Picker

hoodie committed:
Add stable id support to picker delegate

Implement stable id matching for project picker to enable consistent
selection and navigation.

hoodie commented:
https://github.com/user-attachments/assets/0f7fc7c2-5ca0-44cb-8620-f62161a7c00d

I've added it for the recent projects picker too

hoodie committed:
add missing implementations

hoodie committed:
Add stable id support for command palette

hoodie commented:
cool

SomeoneToIgnore committed:
Remove a redundant trait

SomeoneToIgnore committed:
Merge branch 'main' into feature/stabil-picker-selection

SomeoneToIgnore committed:
Post-merge fixes

SomeoneToIgnore committed:
Remove redundant trait bounds

SomeoneToIgnore committed:
Properly handle a few more ids

SomeoneToIgnore committed:
Even less intrisive stable id concept

SomeoneToIgnore committed:
Tidy up

SomeoneToIgnore committed:
Use a better name

SomeoneToIgnore committed:
On confirmation, set the item as manually selected

SomeoneToIgnore reviewed (commented):
I have trimmed it down to the scope where only the pickers influenced by the change have their code adjusted.

-----------------------

When re-testing, I've found 2 things with the selection counter-intuitive: when I select an item manually and first get a query that mismatches the manually selected item entirely (this it's not on the list anymore), it would be good get the manually selected item re-selected on backspace

Now, it seems that backspacing leaves the selection in some odd places sometimes, even without this fallback?
Would be good to keep it closer to top by default, as we try to do now after recent PRs.
